libr/bin/format/mdmp/mdmp.c

```diff
@@ -8,6 +8,122 @@ ut64 r_bin_mdmp_get_baddr(struct r_bin_mdmp_obj *obj) {
 	return (ut64)(obj->b->buf);
 }
 
+ut64 r_bin_mdmp_get_paddr(struct r_bin_mdmp_obj *obj, ut64 vaddr) {
+	/* FIXME: Will only resolve exact matches, probably no need to fix as
+	** this function will become redundant on the optimisation stage */
+	struct minidump_memory_descriptor64 *memory;
+	ut64 index, paddr = 0;
+	RListIter *it;
+
+	/* Loop through the memories sections looking for a match */
+	index = obj->streams.memories64.base_rva;
+	r_list_foreach (obj->streams.memories64.memories, it, memory) {
+		if (vaddr == memory->start_of_memory_range) {
+			paddr = index;
+			break;
+		}
+		index += memory->data_size;
+	}
+	return paddr;
+}
+
+struct minidump_memory_info *r_bin_mdmp_get_mem_info(struct r_bin_mdmp_obj *obj, ut64 vaddr) {
+	struct minidump_memory_info *mem_info;
+	RListIter *it;
+
+	r_list_foreach (obj->streams.memory_infos, it, mem_info) {
+		if (mem_info->allocation_base && vaddr == mem_info->base_address) {
+			return mem_info;
+		}
+	}
+
+	return NULL;
+}
+
+ut32 r_bin_mdmp_get_srwx(struct r_bin_mdmp_obj *obj, ut64 vaddr)
+{
+	bool found = false;
+	struct minidump_memory_info *mem_info;
+	RListIter *it;
+
+	r_list_foreach (obj->streams.memory_infos, it, mem_info) {
+		if (mem_info->allocation_base && vaddr == mem_info->base_address) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) return 0;
+
+	/* FIXME: Have I got these mappings right, I am not sure I have!!! */
+
+	switch (mem_info->protect) {
+	case MINIDUMP_PAGE_READONLY:
+		return R_BIN_SCN_READABLE;
+	case MINIDUMP_PAGE_READWRITE:
+		return R_BIN_SCN_READABLE | R_BIN_SCN_WRITABLE;
+	case MINIDUMP_PAGE_EXECUTE:
+		return R_BIN_SCN_EXECUTABLE;
+	case MINIDUMP_PAGE_EXECUTE_READ:
+		return R_BIN_SCN_EXECUTABLE | R_BIN_SCN_READABLE;
+	case MINIDUMP_PAGE_EXECUTE_READWRITE:
+		return R_BIN_SCN_EXECUTABLE | R_BIN_SCN_READABLE | R_BIN_SCN_WRITABLE;
+	case MINIDUMP_PAGE_NOACCESS:
+	case MINIDUMP_PAGE_WRITECOPY:
+	case MINIDUMP_PAGE_EXECUTE_WRITECOPY:
+	case MINIDUMP_PAGE_GUARD:
+	case MINIDUMP_PAGE_NOCACHE:
+	case MINIDUMP_PAGE_WRITECOMBINE:
+	default:
+		return 0;
+	}
+}
+
+static void r_bin_mdmp_free_pe32_bin(void *pe_bin_) {
+	struct Pe32_r_bin_mdmp_pe_bin *pe_bin = pe_bin_;
+	if (pe_bin) {
+		Pe32_r_bin_pe_free (pe_bin->bin);
+		R_FREE (pe_bin);
+	}
+}
+
+static void r_bin_mdmp_free_pe64_bin(void *pe_bin_) {
+	struct Pe64_r_bin_mdmp_pe_bin *pe_bin = pe_bin_;
+	if (pe_bin) {
+		Pe64_r_bin_pe_free (pe_bin->bin);
+		R_FREE (pe_bin);
+	}
+}
+
+void r_bin_mdmp_free(struct r_bin_mdmp_obj *obj) {
+	if (!obj) return;
+
+	if (obj->streams.ex_threads) r_list_free (obj->streams.ex_threads);
```

Comment:
its unnecessary to if() before freeing, because this check is already done in there